libsolidity/interface/CompilerStack.cpp

```diff
@@ -194,7 +217,30 @@ bool CompilerStack::parse()
 				else
 					noErrors = false;
 
-				m_contracts[contract->name()].contract = contract;
+				if (m_contracts.find(contract->fullyQualifiedName()) != m_contracts.end())
+				{
+					const ContractDefinition* existingContract = m_contracts.find(contract->fullyQualifiedName())->second.contract;
+
+					if (contract != existingContract)
+					{
+						auto err = make_shared<Error>(Error::Type::DeclarationError);
```

Comment:
ContractsCollision1.sol:
```
pragma solidity ^0.4.0;

contract Deed {
}

contract Registrar {
    function newBid(bytes32 sealedBid) payable {
        Deed newBid = new Deed();
    } 
}
```
ContractsCollision2.sol: 
```
pragma solidity ^0.4.0;


contract Deed {
}

contract Registrar {
}

contract NoCollision {
}
```

If you drop something like this right before the if condition we're discussing:

```

          }
          if (contract != existingContract)
          {
            auto err = make_shared<Error>(Error::Type::DeclarationError);
 ```

Then you should see each contract listed once.  It happened so regularly to me that I just assumed it was part of some two-phase compilation process.